agent/src/agentclient.rs

```diff
@@ -0,0 +1,306 @@
+use chrono::naive::NaiveDateTime;
+use k8s_openapi::api::core::v1::Node;
+use kube::api::ListParams;
+use kube::Api;
+use snafu::OptionExt;
+use snafu::ResultExt;
+use std::env;
+use std::process::Command;
+use tokio::time::{sleep, Duration};
+
+use crate::apiclient::{boot_update, get_os_info, list_available, prepare, update};
+use crate::error::{self, Result};
+use apiserver::api::{CreateBottlerocketNodeRequest, UpdateBottlerocketNodeRequest};
+use models::constants::{APISERVER_SERVICE_NAME, APISERVER_SERVICE_PORT, NAMESPACE};
+use models::node::{
+    BottlerocketNode, BottlerocketNodeSelector, BottlerocketNodeState, BottlerocketNodeStatus,
+};
+
+const AGENT_SLEEP_DURATION: Duration = Duration::from_millis(5000);
+
+#[derive(Clone)]
+pub struct BrupopAgent {
+    k8s_client: kube::client::Client,
+}
+
+impl BrupopAgent {
+    pub fn new(k8s_client: kube::client::Client) -> Self {
+        BrupopAgent { k8s_client }
+    }
+
+    pub async fn ensure_node_custom_resource_exists(&self) -> Result<bool> {
+        let associated_bottlerocketnode_name =
+            format!("brn-{}", self.get_node_selector().await?.node_name);
+        let bottlerocket_nodes: Api<BottlerocketNode> =
+            Api::namespaced(self.k8s_client.clone(), NAMESPACE);
+
+        let bottlerocket_nodes_objectlist = bottlerocket_nodes
+            .list(&ListParams::default())
+            .await
+            .context(error::ConvertToObjectList {})?;
+
+        if bottlerocket_nodes_objectlist
+            .iter()
+            .any(|bottlerocket_node_name| {
+                bottlerocket_node_name.metadata.name.as_ref().unwrap()
+                    == &associated_bottlerocketnode_name
+            })
+        {
+            return Ok(true);
+        };
+
+        Ok(false)
```

Comment:
The `unwrap` call will panic if `name` is an `Err<>`, maybe your can try this, but you need to import `ResultExt` from snafu:

```suggestion
        let exists = bottlerocket_nodes_objectlist
            .iter()
            .any(|bottlerocket_node_name| {
                bottlerocket_node_name.metadata.name.as_ref().context(error::MissingName)?
                    == &associated_bottlerocketnode_name
            })

        Ok(exists)
```

Reply:
The name should always exist, assuming the object is being returned by the k8s API. Interestingly, `kube-rs` provides the [`ResourceExt` trait](https://docs.rs/kube/0.60.0/kube/trait.ResourceExt.html#tymethod.name) that you can import to get a method which gets the name, and panics if it doesn't exist, which I think is pretty indicative that the name particularly is rarely missing.

That said, I do like to hedge our bets -- best to avoid panicing if we can!

Reply:
I agree with we should avoid panic if we can, so I would try with arnaldo's suggestion. thanks!

Reply:
I update this to following format because of error `cannot use the ? operator in a closure that returns bool`
```
        for bottlerocket_node in bottlerocket_nodes_objectlist.iter() {
            let bottlerocket_node_name = bottlerocket_node
                .metadata
                .name
                .as_ref()
                .context(error::MissingNodeName)?;
            if bottlerocket_node_name == &associated_bottlerocketnode_name {
                return Ok(true);
            }
        }
```